### PR TITLE
Fix counting of mutations in genome.mutate

### LINKED ISSUE #156: Broken counter for successful mutations in `Genome.mutate()`
I believe there is a problem with counting the mutations in `Genome.mutate() `
The counter `successful_mutations` increases only for mutations in the output region (see genome.py L400, L395: https://github.com/Happy-Algorithms-League/hal-cgp/blob/master/cgp/genome.py#L400). This causes the mutation_factor to be multiplied by a factor ≃ n_hidden_genes / n_output_genes (a bit lower in practice, since some hidden genes might be mutated several times)

Suggestion 1: add `successful_mutations += 1` on line 400. But this causes the new issue, that also mutations that result in the same gene value are counted as successful mutations.
Suggestion 2: add a second output to `_mutate_hidden_region` that indicates whether there was a real change in the value of the gene (success?) and count only if a real mutation happened. 

(PS. point 2 might also be necessary in `_mutate_output_region` as additional control for a "real" mutation)

### PR DESCRIPTION
**Pull request by HenrikMettler**
This PR fixes the mutation counter by explicitly comparing the number of differences in dna before and after mutations

fixes #156

### CODE REVIEW

File: test/test_genome.py [hunk 1]
```diff
@@ -404,6 +404,29 @@ def test_mutate_output_region(rng_seed):
...
+    primitives = (cgp.Add, cgp.Sub, cgp.Mul, cgp.Div, cgp.ConstantFloat)
+
+    rng = np.random.RandomState(rng_seed)
+    genome = cgp.Genome(n_inputs, n_outputs, n_columns, n_rows, levels_back, primitives)
+    genome.randomize(rng)
+
+    dna_diff = 0
+    genome_new = genome.clone()
+    genome_new.mutate(mutation_rate, rng)
+    for idx_dna in range(len(genome.dna)):
```
Comment (jakobj) [inline test/test_genome.py:428]: here you could also `zip` the dnas of the two genomes instead of using the C-style iteration

File: test/test_genome.py [hunk 2]
```diff
@@ -404,6 +404,29 @@ def test_mutate_output_region(rng_seed):
...
+        if genome.dna[idx_dna] != genome_new.dna[idx_dna]:
+            dna_diff += 1
+
+    assert dna_diff == int(mutation_rate * len(genome.dna))
```
Comment (jakobj) [inline test/test_genome.py:432]:
```suggestion
    n_mutations_expected = int(mutation_rate * len(genome.dna))
    assert dna_diff == n_mutations_expected
```
just to make this bit a bit easier to read

File: test/test_genome.py [hunk 3]
```diff
@@ -404,6 +404,29 @@ def test_mutate_output_region(rng_seed):
...
+    n_columns = 10
+    n_rows = 2
+    levels_back = 5
+    primitives = (cgp.Add, cgp.Sub, cgp.Mul, cgp.Div, cgp.ConstantFloat)
+
+    rng = np.random.RandomState(rng_seed)
+    genome = cgp.Genome(n_inputs, n_outputs, n_columns, n_rows, levels_back, primitives)
+    genome.randomize(rng)
+
+    dna_diff = 0
```
Comment (jakobj) [inline test/test_genome.py:425]: i would prefer to call this variable `n_mutations` because it counts the number of mutations

**Comment by jakobj** [comment]
thanks for the update :sparkles: unfortunately travis is unhappy with the new changes, could you please have a look and fix the issues?

File: cgp/genome.py [hunk 4]
```diff
@@ -361,6 +361,14 @@ def mutate(self, mutation_rate: float, rng: np.random.RandomState):
         bool
             True if only inactive regions of the genome were mutated, False otherwise.
         """
+
+        def count_dna_differences(dna_0: List[int], dna_1: List[int]) -> int:
+            dna_diff = 0
+            for gene_0, gene_1 in zip(dna_0, dna_1):
+                if gene_0 != gene_1:
+                    dna_diff += 1
+            return dna_diff
```
Comment (mschmidt87) [inline cgp/genome.py:370]: This can be written as a one-liner:

Comment (mschmidt87) [reply cgp/genome.py:370]:
```suggestion
        def count_dna_differences(dna0: List[int], dna1: List[int]) -> int:
            return len([1 for gene0, gene1 in zip(dna0, dna1) if gene0 != gene1])
```

Comment (mschmidt87) [reply cgp/genome.py:370]: This is just a personal preference (I think), but I'd prefer appending numbers without a `_` to variable names, i.e. to write `dna0, dna1, gene0, gene1` in this case.

File: test/test_genome.py [hunk 5]
```diff
@@ -404,6 +404,30 @@ def test_mutate_output_region(rng_seed):
...
+    rng = np.random.RandomState(rng_seed)
+    genome = cgp.Genome(n_inputs, n_outputs, n_columns, n_rows, levels_back, primitives)
+    genome.randomize(rng)
+
+    n_mutations = 0
+    genome_new = genome.clone()
+    genome_new.mutate(mutation_rate, rng)
+    for (gene_0, gene_1) in zip(genome.dna, genome_new.dna):
+        if gene_0 != gene_1:
+            n_mutations += 1
```
Comment (mschmidt87) [inline test/test_genome.py:430]:
```suggestion
    n_mutations = len([1 for gene0, gene1 in zip(genome.dna, genome_new.dna) if gene0 != gene1])
```

Comment (mschmidt87) [reply test/test_genome.py:430]: You could also import `cgp.genome.count_dna_differences` here.

File: test/test_genome.py [hunk 6]
```diff
@@ -404,6 +404,30 @@ def test_mutate_output_region(rng_seed):
...
 
-    assert genome._mutate_output_region(list(dna), 9, rng) is False
-    assert genome._mutate_output_region(list(dna), 10, rng) is True
-    assert genome._mutate_output_region(list(dna), 11, rng) is False
+    assert genome._mutate_output_region(list(dna), 9, rng) is True
+    assert genome._mutate_output_region(list(dna), 10, rng) is False
+    assert genome._mutate_output_region(list(dna), 11, rng) is True
+
+
+def test_correct_number_of_mutations(mutation_rate, rng_seed):
```
Comment (mschmidt87) [inline test/test_genome.py:412]: I think it would be better to test this with e.g. three mutation rates, instead of just one, using `parametrize` fixture of `pytest`.

File: cgp/genome.py (hunk 4 above)
Comment (jakobj) [reply cgp/genome.py:370]: regarding the one liner: please show both to your colleague for 5sec and ask them if the understand what's happening ;) i do like compact code but it feels like this change would be sacrificing readability

File: test/test_genome.py (hunk 5 above)
Comment (jakobj) [reply test/test_genome.py:430]: i would not reuse the function from the library, the probability of implementing it wrong two times is trivially smaller than implementing it wrong once ;)

**Comment by jakobj** [comment]
regarding the example: the mutation rate was chosen accordingly to the CGP literature, assuming the implementation is correct, so it should be fine. seems like they all still work, don't they?

File: test/test_genome.py (hunk 6 above)
Comment (HenrikMettler) [reply test/test_genome.py:413]: In the newest commit I added: `@pytest.mark.parametrize("mutation_rate", [0.02, 0.05, 0.2,])` just before this test. Is that the right way to do it?

File: cgp/genome.py (hunk 4 above)
Comment (mschmidt87) [reply cgp/genome.py]: Hmm, I understand your point in general, but in this case I think the list comprehension is pretty readable and more readable than the original code.

File: test/test_genome.py (hunk 5 above)
Comment (mschmidt87) [reply test/test_genome.py]: > regarding the example: the mutation rate was chosen accordingly to the CGP literature, assuming the implementation is correct, so it should be fine. seems like they all still work, don't they?

I don't know, I haven't run them. The CI only checks whether they crash, but not if they convert to good fitness values.

**Comment by mschmidt87** [comment]
> regarding the example: the mutation rate was chosen accordingly to the CGP literature, assuming the implementation is correct, so it should be fine. seems like they all still work, don't they?

I haven't tested the examples. The CI only checks if they crash, not if they convert.

File: test/test_genome.py (hunk 6 above)
Comment (mschmidt87) [reply test/test_genome.py:413]: Looks good to me, although I am not sure whether this might interfer with the fixture from `conftest.py` that has the same name.

File: test/test_genome.py (hunk 5 above)
Comment (mschmidt87) [reply test/test_genome.py]: Okay for me.

File: cgp/genome.py (hunk 4 above)
Comment (jakobj) [reply cgp/genome.py]: ok

**Comment by mschmidt87** [comment]
I think that CI is failing because of `flake8` issues. I would recommend to install pre-commit hooks for git to catch such errors locally, @HenrikMettler .

File: test/test_genome.py [hunk 7]
```diff
@@ -404,6 +404,29 @@ def test_mutate_output_region(rng_seed):
...
+    primitives = (cgp.Add, cgp.Sub, cgp.Mul, cgp.Div, cgp.ConstantFloat)
+
+    rng = np.random.RandomState(rng_seed)
+    genome = cgp.Genome(n_inputs, n_outputs, n_columns, n_rows, levels_back, primitives)
+    genome.randomize(rng)
+
+    n_mutations = 0
+    genome_new = genome.clone()
+    genome_new.mutate(mutation_rate, rng)
+    n_mutations = len([1 for gene0, gene1 in zip(genome.dna, genome_new.dna) if gene0 != gene1])
```
Comment (jakobj) [inline test/test_genome.py:429]: even under the assumption of the one liner being acceptable in the main module, i would strongly suggest to use the long, slow, but super easy to write and read version in the tests. test should always be very verbose, cool tricks can be applied in the main module.

**Comment by jakobj** [comment]
> > regarding the example: the mutation rate was chosen accordingly to the CGP literature, assuming the implementation is correct, so it should be fine. seems like they all still work, don't they?
> 
> I haven't tested the examples. The CI only checks if they crash, not if they convert.

good point! @HenrikMettler please also manually test the examples and check whether they converge properly.

File: test/test_genome.py (hunk 7 above)
Comment (HenrikMettler) [reply test/test_genome.py:429]: my bad I didn't notice your reply to the above comments, sorry

Comment (jakobj) [reply test/test_genome.py:429]: no worries, you can just "unfix" it again in the test ;)

**Comment by HenrikMettler** [comment]
For the examples: 
- all of them stop prior to the max number of generations with values ~ -1e-15 (to -1e-30) -> that this is convergence I assume? 
- `example_evo_regression.py`: The first only converges after 950 of (max) 1000 generations. Should I increase the max number of allowed generations?

**Comment by jakobj** [comment]
no, i think it's fine, no need to change it as long as it converges within the given number of generations. thanks for checking this :)

**Comment by HenrikMettler** [comment]
the latest commit should clear all the issues. I can't squash though (I think) since i don't have the rights to squash and merge (if I understand this: [https://cloudfour.com/thinks/squashing-your-pull-requests/](url) correctly)

File: examples/example_differential_evo_regression.py [hunk 8]
```diff
@@ -126,7 +126,7 @@ def recording_callback(pop):
 
     history, champion = evolution()
 
-    print(f"Final expression {champion.to_sympy()[0]} with fitness {champion.fitness}")
+    print(f"Final expression {champion.to_sympy()[0]} " f"with fitness {champion.fitness}")
```
Comment (jakobj) [inline examples/example_differential_evo_regression.py:129]: why this change?

Comment (HenrikMettler) [reply examples/example_differential_evo_regression.py:129]: hmm... i had to do a linebreak to make `flake` happy... and then I guess `black` removed it again?

Comment (jakobj) [reply examples/example_differential_evo_regression.py:129]: that's strange. could you try removing the `" f"` again and see what travis thinks? maybe after all you need to run flake with `flake8 --config=.flake8 .` in contrast to what i might've said :P

**Comment by mschmidt87** [comment]
Before we merge, one more (minor) thing: could you make the commit message a bit nicer, i.e., remove the "some more changes" lines etc., and stick to imperative tense and capitalized beginning?
